tests/src/gui/testqgsattributeform.cpp

```diff
@@ -906,6 +907,107 @@ void TestQgsAttributeForm::testAttributeFormInterface()
 }
 
 
+void TestQgsAttributeForm::testDefaultValueUpdate()
+{
+  // make a temporary layer to check through
+  QString def = QStringLiteral( "Point?field=col0:integer&field=col1:integer&field=col2:integer&field=col3:integer" );
+  QgsVectorLayer *layer = new QgsVectorLayer( def, QStringLiteral( "test" ), QStringLiteral( "memory" ) );
+
+  //set defaultValueDefinitions
+  //col0 - no default value
+  //col1 - "col0"+1
+  //col2 - "col0"+"col1"
+  //col3 - "col2"
+
+  // set constraints for each field
+  layer->setDefaultValueDefinition( 1, QgsDefaultValue( QStringLiteral( "\"col0\"+1" ) ) );
+  layer->setDefaultValueDefinition( 2, QgsDefaultValue( QStringLiteral( "\"col0\"+\"col1\"" ) ) );
+  layer->setDefaultValueDefinition( 3, QgsDefaultValue( QStringLiteral( "\"col2\"" ) ) );
+
+  // build a form for this feature
+  QgsFeature ft( layer->dataProvider()->fields(), 1 );
+  ft.setAttribute( QStringLiteral( "col0" ), 0 );
+  QgsAttributeForm form( layer );
+  form.setMode( QgsAttributeEditorContext::AddFeatureMode );
+  form.setFeature( ft );
+
+  // get wrappers for each widget
+  QgsEditorWidgetWrapper *ww0, *ww1, *ww2, *ww3;
+  ww0 = qobject_cast<QgsEditorWidgetWrapper *>( form.mWidgets[0] );
+  ww1 = qobject_cast<QgsEditorWidgetWrapper *>( form.mWidgets[1] );
+  ww2 = qobject_cast<QgsEditorWidgetWrapper *>( form.mWidgets[2] );
+  ww3 = qobject_cast<QgsEditorWidgetWrapper *>( form.mWidgets[3] );
+
+  //set value in col0:
+  ww0->setValue( 5 );
+
+  //we expect
+  //col0 - 5
+  //col1 - 6
+  //col2 - 11
+  //col3 - 11
+
+  QCOMPARE( ww0->value().toInt(), 5 );
+  QCOMPARE( ww1->value().toInt(), 6 );
+  QCOMPARE( ww2->value().toInt(), 11 );
+  QCOMPARE( ww3->value().toInt(), 11 );
+
+  //set value in col1:
+  ww1->setValue( 10 );
+
+  //we expect
+  //col0 - 5
+  //col1 - 10
+  //col2 - 15
+  //col3 - 15
+
+  QCOMPARE( ww0->value().toInt(), 5 );
+  QCOMPARE( ww1->value().toInt(), 10 );
+  QCOMPARE( ww2->value().toInt(), 15 );
+  QCOMPARE( ww3->value().toInt(), 15 );
+
+  /*
```

Comment:
What's up with this part of the code?

Reply:
Preparations for my later commit to test if there is a recursion.